content/docs/devops-tips/scaling-cert-manager.md

```diff
@@ -0,0 +1,134 @@
+> the Certificate's history. Each revision represents a single
+> `CertificateRequest` created by this Certificate, either when it was
+> created, renewed, or Spec was changed. Revisions will be removed by oldest
+> first if the number of revisions exceeds this number.
+>  If set, `revisionHistoryLimit` must be a value of `1` or greater. If unset
+> (`nil`), revisions will not be garbage collected. Default value is `nil`.
+
+On a busy cluster these will eventually overwhelm your Kubernetes API server;
+because of the memory and CPU required to cache them all and the storage required to save them.
+
+Use a tool like Kyverno to override the `Certificate.spec.revisionHistoryLimit` for all namespaces.
+
+> 📖 Adapt [the Kyverno policies in the tutorial: how to set Certificate defaults automatically](../tutorials/certificate-defaults/README.md),
+> to override rather than default the `revisionHistoryLimit` field.
+>
+> 📖 Learn [how to set `revisionHistoryLimit` when using Annotated Ingress resources](../usage/ingress.md#supported-annotations).
+>
+> 🔗 Read [`cert-manager#3773`: Certificate revision history limit](https://github.com/cert-manager/cert-manager/pull/3773),
+> to learn why stale CertificateRequests resources are not automatically deleted.
+
+## Enable Server-Side Apply
+
+By default, cert-manager [uses Update requests](https://kubernetes.io/docs/reference/using-api/api-concepts/#update-mechanism-update)
+to create and modify resources like CertificateRequest and Secret,
+but on a busy cluster there will be frequent conflicts as the control loops in cert-manager each try to update the status of various resources.
+
+You will see errors, like this one, in the logs:
+
+> `I0419 14:11:51.325377       1 controller.go:162] "re-queuing item due to optimistic locking on resource" logger="cert-manager.certificates-trigger" key="team-864-p6ts6/app-7" error="Operation cannot be fulfilled on certificates.cert-manager.io \"app-7\": the object has been modified; please apply your changes to the latest version and try again"`
+
+This error is relatively harmless because the update attempt is retried,
+but it slows down the reconciliation because each error triggers an exponential back off mechanism,
+which causes increasing delays between retries.
+
+The solution is to turn on the [Server-Side Apply Feature](../installation/configuring-components.md#feature-gates),
+which causes cert-manager to use [HTTP PATCH using Server-Side Apply](https://kubernetes.io/docs/reference/using-api/api-concepts/#update-mechanism-server-side-apply) when ever it needs to modify an API resource.
+This avoids all conflicts because each cert-manager controller sets only the fields that it owns.
+
+You can enable the server-side apply feature gate with the following Helm chart values:
+
+```yaml
+# helm-values.yaml
+config:
+  apiVersion: controller.config.cert-manager.io/v1alpha1
+  kind: ControllerConfiguration
+  featureGates:
+    AllBeta: true
```

Comment:
Isn't this the default? Do we have to specifically enable `AllBeta` here?

Reply:
I think so, but I cannot find the exact place in the code that says as such.
It can be good to be explicit in whats enable with helm values, but it's a preference choice.

@wallrj - assuming this is default, I'd drop that flag.. OR just add a comment that this is the default for reference.

Reply:
Agreed. Dropped. I think I assumed it necessary, because I saw it in the E2E setup scripts:

https://github.com/cert-manager/cert-manager/blob/5d0b738b3e35679f1fec3b7a5deac4e7f5a48050/make/e2e-setup.mk#L263-L267